monitoring/aws.go

```diff
@@ -38,26 +39,46 @@ type awsHasProfileChecker struct{}
 
 // Name returns this checker name
 // Implements health.Checker
-func (*awsHasProfileChecker) Name() string {
+func (c *awsHasProfileChecker) Name() string {
 	return awsHasProfileCheckerID
 }
 
 // Check will check the metadata API to see if an IAM profile is assigned to the node
 // Implements health.Checker
-func (*awsHasProfileChecker) Check(ctx context.Context, reporter health.Reporter) {
+func (c *awsHasProfileChecker) Check(ctx context.Context, reporter health.Reporter) {
+	probesCh := make(chan health.Reporter, 1)
+	go func() {
+		probes, err := c.check()
+		if err != nil {
+			probes.Add(NewProbeFromErr(c.Name(), "failed to validate IAM profile", err))
+		}
+		probesCh <- probes
+	}()
+	select {
+	case probes := <-probesCh:
```

Comment:
This is still backwards to me: the goroutine would rather handle the context and you'd wait for results (otherwise the goroutine is leaked for the duration of the request):
```go
func (c *awsHasProfileChecker) Check(ctx context.Context, reporter health.Reporter) {
	probesCh := make(chan health.Reporter, 1)
	go func() {
		probes, err := c.check(ctx)
		if err != nil {
			probes.Add(NewProbeFromErr(c.Name(), "failed to validate IAM profile", err))
		}
		probesCh <- probes
	}()
	probes := <-probesCh
	health.AddFrom(reporter, probes)
}
```
Now, it is not quite possible with the defaults (i.e. instance of EC2metadata does not use context in its APIs) but with some extra work - for example, by using a [Waiter](https://github.com/aws/aws-sdk-go/blob/v1.13.57/aws/request/waiter.go#L74):
```go
// check will check the metadata API to see if an IAM profile is assigned to the node.
func (c *awsHasProfileChecker) check(ctx context.Context) (probes health.Reporter, err error) {
	probes = &health.Probes{}

	session, err := session.NewSession()
	if err != nil {
		return probes, trace.Wrap(err, "failed to create session")
	}

	metadata := ec2metadata.New(session)
	err = c.iamInfo(ctx, metadata)
	if err != nil {
		return probes, trace.Wrap(err, "failed to determine node IAM profile")
	}

	probes.Add(NewSuccessProbe(c.Name()))
	return probes, nil
}

func (c *awsHasProfileChecker) iamInfo(ctx context.Context, metadata *ec2netadata.EC2Metadata) error {
	resp, err := c.getIAMInfo(ctx, metadata)
	if err != nil {
		return trace.Wrap(err, "failed to get EC2 IAM info")
	}

	var info ec2metadata.EC2IAMInfo
	if err := json.NewDecoder(strings.NewReader(resp)).Decode(&info); err != nil {
		return trace.Wrap(err, "failed to decode EC2 IAM info")
	}

	if info.Code != "Success" {
		return trace.Wrap(err, "failed to get EC2 IAM Info (%s)", info.Code)
	}
	return nil
}

func (c *awsHasProfileChecker) getIAMInfo(ctx context.Context, metadata *ec2metadata.EC2Metadata) error {
	waiter := &request.Waiter{
		// other fields skipped
		NewRequest: func([]request.Option) (*request.Request, error) {
			op := &request.Operation{
				Name:       "GetMetadata",
				HTTPMethod: "GET",
				HTTPPath:   path.Join("/", "meta-data", "iam/info"),
			}
			req := metadata.NewRequest(op, nil, nil)
			req.SetContext(ctx)
			return req
		},
	}
	return waiter.WaitWithContext(ctx)
}
```
The only issue with this approach, of course, is that it "leaks" some of the internal knowledge (`iam-info` endpoint).
Maybe there's a better way though...

Reply:
```
func (c *awsHasProfileChecker) Check(ctx context.Context, reporter health.Reporter) {
	probesCh := make(chan health.Reporter, 1)
	go func() {
		probes, err := c.check(ctx)
		if err != nil {
			probes.Add(NewProbeFromErr(c.Name(), "failed to validate IAM profile", err))
		}
		probesCh <- probes
	}()
	probes := <-probesCh
	health.AddFrom(reporter, probes)
}
```
If we pass ctx into check then is there any need for this goroutine?
Could we just change this to:
```
func (c *awsHasProfileChecker) Check(ctx context.Context, reporter health.Reporter) {
	probes, err := c.check(ctx)
	if err != nil {
		probes.Add(NewProbeFromErr(c.Name(), "failed to validate IAM profile", err))
	}
	health.AddFrom(reporter, probes)
}
```

Reply:
Sure - this is even better.